massa-grpc/src/public.rs

```
@@ -0,0 +1,735 @@
+            }
+
+            Some(grpc_model::OperationWrapper {
+                id: secure_share.id.to_string(),
+                thread: secure_share
+                    .content_creator_address
+                    .get_thread(grpc.grpc_config.thread_count) as u32,
+                operation: Some((*secure_share).clone().into()),
+                block_ids: block_ids.into_iter().map(|id| id.to_string()).collect(),
+            })
+        })
+        .collect();
+
+    Ok(grpc_api::GetOperationsResponse {
+        wrapped_operations: operations,
+    })
+}
+
+/// Get smart contract execution events
+pub(crate) fn get_sc_execution_events(
+    grpc: &MassaPublicGrpc,
+    request: tonic::Request<grpc_api::GetScExecutionEventsRequest>,
+) -> Result<grpc_api::GetScExecutionEventsResponse, GrpcError> {
+    let event_filter = to_event_filter(request.into_inner().filters)?;
+    let events: Vec<grpc_model::ScExecutionEvent> = grpc
+        .execution_controller
+        .get_filtered_sc_output_event(event_filter)
+        .into_iter()
+        .map(|event| event.into())
+        .collect();
+
+    Ok(grpc_api::GetScExecutionEventsResponse { events })
+}
+
+//  Get selector draws
+pub(crate) fn get_selector_draws(
+    grpc: &MassaPublicGrpc,
+    request: tonic::Request<grpc_api::GetSelectorDrawsRequest>,
+) -> Result<grpc_api::GetSelectorDrawsResponse, GrpcError> {
+    let inner_req = request.into_inner();
+    let mut addresses: PreHashSet<Address> = PreHashSet::default();
+    let mut slot_range: (Option<Slot>, Option<Slot>) = (None, None);
+
+    // parse filters from request
+    inner_req.filters.into_iter().for_each(|query| {
+        if let Some(filter) = query.filter {
+            match filter {
+                grpc_api::selector_draws_filter::Filter::Addresses(addrs) => {
+                    addrs
+                        .addresses
+                        .into_iter()
+                        .for_each(|addr| match Address::from_str(&addr) {
```

@modship IMHO we should fail fast here